### ISSUE TITLE
significant_drop_tightening false positive on loops

### ISSUE THREAD

**Issue by Snowiiii** [rust-lang/rust-clippy#15510]
### Summary

I want to lock my RwLock before the Loop so i don't have to lock it on every iteration saving performance
```rust
let mut entities = world.entities.write().await;
for (uuid, entity_nbt) in &chunk.read().await.data {
     ...
     entities.remove(&base_entity.entity_uuid);
}
```
when having `significant_drop_tightening` enabled it suggests me to use this, inside the loop
```rust
world.entities.write().await.remove(&base_entity.entity_uuid);
```

### Lint Name

significant_drop_tightening

### Reproducer

I tried this code:

```rust
let mut entities = world.entities.write().await;
for (uuid, entity_nbt) in &chunk.read().await.data {
     ...
     entities.remove(&base_entity.entity_uuid);
}
```

I saw this happen:

```rust
world.entities.write().await.remove(&base_entity.entity_uuid);
```

I expected to see this happen:

### Version

```text
1.89
```

**Comment by KisaragiEffective** [comment]
I think triggering lint on it is *logically* correct so I-false-positive wouldn't fit on your case?
Also, your snippet isn't self-contained; it makes hard to reproduce it because [there are so many candidate types](https://github.com/search?q=language%3Arust+%2Fclippy%3A%3Ahas_significant_drop%2F+NOT+is%3Afork&type=code).

P.S. It would be some `RwLock`s, not `Mutex`; latter does not have method that `read()` and `write()`.
